File: src/geo/atmosphere-engine.ts
// ─── Atmosphere Engine ───────────────────────────────────────────────────────
// Phase 4 orchestrator: runs ClimateEngine → WeatherEngine pipeline,
// emits climate/weather events, handles sub-tick batching.

import type { StateBufferViews, AtmosphericComposition } from '../shared/types';
import { GRID_SIZE } from '../shared/types';
import type { EventBus } from '../kernel/event-bus';
import type { EventLog } from '../kernel/event-log';
import { ClimateEngine, type ClimateResult } from './climate-engine';
import { WeatherEngine, type WeatherResult } from './weather-engine';
import { Xoshiro256ss } from '../proc/prng';

// ─── Types ──────────────────────────────────────────────────────────────────

export interface AtmosphereTickResult {
  climate: ClimateResult;
  weather: WeatherResult;
}

export interface AtmosphereEngineConfig {
  /** Minimum atmosphere tick interval in Ma (default 1.0). */
  minTickInterval?: number;
  /** Grid size (default GRID_SIZE). */
  gridSize?: number;
}

// ─── AtmosphereEngine ───────────────────────────────────────────────────────

export class AtmosphereEngine {
  private readonly climateEngine: ClimateEngine;
  private readonly weatherEngine: WeatherEngine;
  private readonly bus: EventBus;
  private readonly eventLog: EventLog;
  private readonly rng: Xoshiro256ss;

  private stateViews: StateBufferViews | null = null;
  private atmosphere: AtmosphericComposition | null = null;

  /** Sub-tick accumulator. */
  private accumulator = 0;
  private readonly minTickInterval: number;

  /** Previous ice-age state for onset/end detection. */
  private prevIceAge = false;

  constructor(
    bus: EventBus,
    eventLog: EventLog,
    seed: number,
    config?: AtmosphereEngineConfig,
  ) {
    const gridSize = config?.gridSize ?? GRID_SIZE;
    this.bus = bus;
    this.eventLog = eventLog;
    this.rng = new Xoshiro256ss(seed);
    this.minTickInterval = config?.minTickInterval ?? 1.0;

    this.climateEngine = new ClimateEngine(gridSize);
    this.weatherEngine = new WeatherEngine(gridSize);
  }

  /**
   * Initialize with shared state and atmospheric composition.
   */
  initialize(
    stateViews: StateBufferViews,
    atmosphere: AtmosphericComposition,
  ): void {
    this.stateViews = stateViews;
    this.atmosphere = atmosphere;
    this.accumulator = 0;
    this.prevIceAge = false;
  }

  /**
   * Process one simulation tick.
   * Batches sub-ticks if deltaMa exceeds the minimum tick interval.
   */
  tick(timeMa: number, deltaMa: number): AtmosphereTickResult | null {
    if (!this.stateViews || !this.atmosphere || deltaMa <= 0) return null;

    this.accumulator += deltaMa;

    let lastResult: AtmosphereTickResult | null = null;

    while (this.accumulator >= this.minTickInterval) {
      const subDelta = this.minTickInterval;
      this.accumulator -= subDelta;
      const subTime = timeMa - this.accumulator;

      lastResult = this.processAtmosphereTick(subTime, subDelta);
    }

    return lastResult;
  }

  // ── Core tick processing ──────────────────────────────────────────────

  private processAtmosphereTick(timeMa: number, deltaMa: number): AtmosphereTickResult {
    const stateViews = this.stateViews!;
    const atmosphere = this.atmosphere!;

    // 1 — Climate (insolation, 3-cell winds, Milankovitch, greenhouse)
    const climate = this.climateEngine.tick(
      timeMa, deltaMa, stateViews, atmosphere, this.rng,
    );

    // 2 — Weather (fronts, cyclones, orographic precipitation, clouds)
    const weather = this.weatherEngine.tick(
      timeMa, deltaMa, stateViews, this.rng,
    );

    // ── Emit events ───────────────────────────────────────────────────────

    // Climate update
    this.bus.emit('CLIMATE_UPDATE', {
      meanTemperature: climate.meanTemperature,
      co2Ppm: climate.co2Ppm,
      iceAlbedoFeedback: climate.iceAlbedoFeedback,
    });

    // Tropical cyclones
    for (const cyclone of weather.tropicalCyclones) {
      this.bus.emit('TROPICAL_CYCLONE_FORMED', {
        lat: cyclone.lat,
        lon: cyclone.lon,
        intensity: cyclone.intensity,
      });
      this.eventLog.record({
        timeMa,
        type: 'TROPICAL_CYCLONE_FORMED',
        description: `Tropical cyclone Cat ${cyclone.intensity} at ${cyclone.lat.toFixed(1)}°, ${cyclone.lon.toFixed(1)}°`,
      });
    }

    // Snowball Earth
    if (climate.snowballTriggered) {
      this.bus.emit('SNOWBALL_EARTH', {
        equatorialTemp: climate.equatorialTemperature,
      });
      this.eventLog.record({
        timeMa,
        type: 'SNOWBALL_EARTH',
        description: `Snowball Earth threshold reached: equatorial temp ${climate.equatorialTemperature.toFixed(1)}°C`,
      });
    }

    // Ice age onset / end based on mean temperature threshold
    const isIceAge = climate.meanTemperature < -5;
    if (isIceAge && !this.prevIceAge) {
      this.bus.emit('ICE_AGE_ONSET', { severity: climate.iceAlbedoFeedback });
      this.eventLog.record({
        timeMa,
        type: 'ICE_AGE_ONSET',
        description: `Ice age onset: mean temp ${climate.meanTemperature.toFixed(1)}°C, ice coverage ${(climate.iceAlbedoFeedback * 100).toFixed(1)}%`,
      });
    } else if (!isIceAge && this.prevIceAge) {
      this.bus.emit('ICE_AGE_END', {});
      this.eventLog.record({
        timeMa,
        type: 'ICE_AGE_END',
        description: `Ice age ended: mean temp recovered to ${climate.meanTemperature.toFixed(1)}°C`,
      });
    }
    this.prevIceAge = isIceAge;

    return { climate, weather };
  }

  // ── Getters ──────────────────────────────────────────────────────────────

  getClimateEngine(): ClimateEngine {
    return this.climateEngine;
  }

  getWeatherEngine(): WeatherEngine {
    return this.weatherEngine;
  }
}

File: backend/GeoTime.Core/Engines/AtmosphereEngine.cs
using GeoTime.Core.Models;
using GeoTime.Core.Kernel;
using GeoTime.Core.Proc;

namespace GeoTime.Core.Engines;

/// <summary>Phase 4 orchestrator: climate + weather pipeline.</summary>
public sealed class AtmosphereEngine(EventBus bus, EventLog log, uint seed, int gridSize, double minTick = 1.0)
{
    private readonly ClimateEngine _climate = new(gridSize);
    private readonly WeatherEngine _weather = new(gridSize);
    private readonly Xoshiro256ss _rng = new(seed);

    private SimulationState? _state;
    private AtmosphericComposition? _atmo;
    private double _accumulator;
    private bool _prevIceAge;

    public void Initialize(SimulationState state, AtmosphericComposition atmo)
    {
        _state = state; _atmo = atmo; _accumulator = 0; _prevIceAge = false;
    }

    public AtmosphereTickResult? Tick(double timeMa, double deltaMa)
    {
        if (_state == null || _atmo == null || deltaMa <= 0) return null;
        _accumulator += deltaMa;
        AtmosphereTickResult? last = null;
        while (_accumulator >= minTick) { _accumulator -= minTick; last = Process(timeMa - _accumulator, minTick); }
        return last;
    }

    private AtmosphereTickResult Process(double t, double dt)
    {
        var cl = _climate.Tick(t, dt, _state!, _atmo!, _rng);
        var we = _weather.Tick(t, dt, _state!, _rng);

        bus.Emit("CLIMATE_UPDATE", new { meanTemperature = cl.MeanTemperature, co2Ppm = cl.CO2Ppm, iceAlbedoFeedback = cl.IceAlbedoFeedback });

        foreach (var c in we.TropicalCyclones)
        {
            bus.Emit("TROPICAL_CYCLONE_FORMED", new { lat = c.Lat, lon = c.Lon, intensity = c.Intensity });
            log.Record(new GeoLogEntry { TimeMa = t, Type = "TROPICAL_CYCLONE_FORMED", Description = $"Cat {c.Intensity} at {c.Lat:F1}°, {c.Lon:F1}°" });
        }

        if (cl.SnowballTriggered)
        {
            bus.Emit("SNOWBALL_EARTH", new { equatorialTemp = cl.EquatorialTemperature });
            log.Record(new GeoLogEntry { TimeMa = t, Type = "SNOWBALL_EARTH", Description = $"Equatorial temp {cl.EquatorialTemperature:F1}°C" });
        }

        var iceAge = cl.MeanTemperature < -5;
        if (iceAge && !_prevIceAge)
        {
            bus.Emit("ICE_AGE_ONSET", new { severity = cl.IceAlbedoFeedback });
            log.Record(new GeoLogEntry { TimeMa = t, Type = "ICE_AGE_ONSET", Description = $"Mean temp {cl.MeanTemperature:F1}°C" });
        }
        else if (!iceAge && _prevIceAge)
        {
            bus.Emit("ICE_AGE_END", new { });
            log.Record(new GeoLogEntry { TimeMa = t, Type = "ICE_AGE_END", Description = $"Mean temp recovered to {cl.MeanTemperature:F1}°C" });
        }
        _prevIceAge = iceAge;

        return new AtmosphereTickResult { Climate = cl, Weather = we };
    }
}

public sealed class AtmosphereTickResult
{
    public ClimateResult Climate { get; set; } = new();
    public WeatherResult Weather { get; set; } = new();
}
